File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/internal/pipeline/TrackTimeline.kt
package com.swmansion.moqkit.subscribe.internal.pipeline

internal enum class TimelineDropReason {
    /** Frame is older than the current playback position by more than the freshness budget. */
    STALE_VS_PLAYBACK,

    /** Transport groups were skipped to keep playback within its live-latency budget. */
    LATENCY_BUDGET_SKIP,

    /** Transport groups were evicted before the timeline could consume them. */
    NETWORK_EVICTED,

    /** Transport groups were superseded by another range covering the same media. */
    COVERED,

    /** Transport groups were skipped because the publisher sequence moved backward. */
    REWIND,

    /** An expected transport sequence range was never delivered. */
    MISSING_SEQUENCE,
}

internal enum class TimelineResetReason {
    /** Publisher started a new epoch after rewinding its media sequence. */
    PUBLISHER_REWIND,

    /** Consecutive media timestamps differ by more than the configured maximum gap. */
    TIMESTAMP_GAP,

    /** A downstream decoder or renderer recovery requested a local timeline reset. */
    DOWNSTREAM_RECOVERY,
}

internal sealed interface TimelineDecision {
    data class Admit(val frame: TimedFrame) : TimelineDecision

    data class Drop(
        val reason: TimelineDropReason,
        val frame: TimedFrame? = null,
        val groupRange: LongRange? = null,
    ) : TimelineDecision

    data class Reset(
        val reason: TimelineResetReason,
        val epoch: Long,
        val resumeFrom: TimedFrame?,
        val gapUs: Long? = null,
    ) : TimelineDecision

    data class End(val error: PipelineError?) : TimelineDecision
}

/**
 * Sole authority for track epoch, freshness, timestamp-gap, and latency decisions.
 * The class is synchronous and deterministic; adapters supply all clock observations.
 */
internal class TrackTimeline(
    private val policy: TimelinePolicy,
    private val timeSource: TimeSource,
) {
    private var liveEdgeOffsetUs: Long? = null
    private var playbackPosition: Long? = null
    private var lastTimestamp: Long? = null

    var currentEpoch: Long? = null
        private set

    var targetLatencyUs: Long = policy.targetLatencyUs
        private set

    @Synchronized
    fun onIngest(event: IngestEvent): TimelineDecision = when (event) {
        is IngestEvent.Frame -> onFrame(event)
        is IngestEvent.GroupsSkipped -> TimelineDecision.Drop(
            reason = event.reason.toTimelineDropReason(),
            groupRange = event.fromSequence..event.toSequence,
        )
        is IngestEvent.Discontinuity -> resetForEpoch(event.epoch, resumeFrom = null)
        is IngestEvent.Closed -> TimelineDecision.End(event.error)
    }

    @Synchronized
    fun onPlaybackPosition(positionUs: Long) {
        require(positionUs >= 0L) { "positionUs must be non-negative" }
        playbackPosition = positionUs
    }

    @Synchronized
    fun requestReset(reason: TimelineResetReason = TimelineResetReason.DOWNSTREAM_RECOVERY): TimelineDecision.Reset {
        lastTimestamp = null
        liveEdgeOffsetUs = null
        playbackPosition = null
        return TimelineDecision.Reset(
            reason = reason,
            epoch = currentEpoch ?: 0L,
            resumeFrom = null,
        )
    }

    @Synchronized
    fun liveEdgeUs(): Long? {
        val offsetUs = liveEdgeOffsetUs ?: return null
        return addOrNull(timeSource.nanoTime() / NANOS_PER_MICROSECOND, offsetUs)
    }

    @Synchronized
    fun currentLatencyUs(): Long? {
        val edge = liveEdgeUs() ?: return null
        val playback = playbackPosition ?: return null
        return subtractClampedToZero(edge, playback)
    }

    @Synchronized
    fun setTargetLatency(us: Long) {
        require(us >= 0L) { "target latency must be non-negative" }
        targetLatencyUs = us
    }

    private fun onFrame(event: IngestEvent.Frame): TimelineDecision {
        val frame = event.frame
        val epoch = currentEpoch
        if (epoch != null && frame.epoch != epoch) {
            return resetForEpoch(frame.epoch, resumeFrom = frame, arrivalNanos = event.arrivalNanos)
        }
        if (epoch == null) {
            currentEpoch = frame.epoch
        }

        val previousTimestamp = lastTimestamp
        val gapUs = previousTimestamp?.let { absoluteDifferenceSaturated(frame.timestampUs, it) }
        if (gapUs != null && gapUs > policy.maxGapUs) {
            lastTimestamp = frame.timestampUs
            resetLiveEdge(frame.timestampUs, event.arrivalNanos)
            playbackPosition = null
            return TimelineDecision.Reset(
                reason = TimelineResetReason.TIMESTAMP_GAP,
                epoch = frame.epoch,
                resumeFrom = frame,
                gapUs = gapUs,
            )
        }

        val playback = playbackPosition
        if (playback != null && isOlderThanFreshnessBudget(frame.timestampUs, playback)) {
            return TimelineDecision.Drop(
                reason = TimelineDropReason.STALE_VS_PLAYBACK,
                frame = frame,
            )
        }

        lastTimestamp = frame.timestampUs
        recordLiveEdge(frame.timestampUs, event.arrivalNanos)
        return TimelineDecision.Admit(frame)
    }

    private fun resetForEpoch(
        epoch: Long,
        resumeFrom: TimedFrame?,
        arrivalNanos: Long? = null,
    ): TimelineDecision.Reset {
        require(epoch >= 0L) { "epoch must be non-negative" }
        currentEpoch = epoch
        lastTimestamp = resumeFrom?.timestampUs
        liveEdgeOffsetUs = null
        playbackPosition = null
        if (resumeFrom != null && arrivalNanos != null) {
            recordLiveEdge(resumeFrom.timestampUs, arrivalNanos)
        }
        return TimelineDecision.Reset(
            reason = TimelineResetReason.PUBLISHER_REWIND,
            epoch = epoch,
            resumeFrom = resumeFrom,
        )
    }

    private fun isOlderThanFreshnessBudget(timestampUs: Long, playbackUs: Long): Boolean {
        if (timestampUs >= playbackUs) return false
        val ageUs = subtractOrNull(playbackUs, timestampUs) ?: Long.MAX_VALUE
        return ageUs > policy.freshnessBudgetUs
    }

    private fun subtractClampedToZero(left: Long, right: Long): Long =
        if (left <= right) 0L else subtractOrNull(left, right) ?: Long.MAX_VALUE

    private fun recordLiveEdge(timestampUs: Long, arrivalNanos: Long) {
        val offsetUs = subtractOrNull(timestampUs, arrivalNanos / NANOS_PER_MICROSECOND) ?: return
        liveEdgeOffsetUs = maxOf(liveEdgeOffsetUs ?: Long.MIN_VALUE, offsetUs)
    }

    private fun resetLiveEdge(timestampUs: Long, arrivalNanos: Long) {
        liveEdgeOffsetUs = null
        recordLiveEdge(timestampUs, arrivalNanos)
    }

    private fun absoluteDifferenceSaturated(left: Long, right: Long): Long =
        if (left >= right) subtractOrNull(left, right) ?: Long.MAX_VALUE
        else subtractOrNull(right, left) ?: Long.MAX_VALUE

    private fun subtractOrNull(left: Long, right: Long): Long? = try {
        Math.subtractExact(left, right)
    } catch (_: ArithmeticException) {
        null
    }

    private fun addOrNull(left: Long, right: Long): Long? = try {
        Math.addExact(left, right)
    } catch (_: ArithmeticException) {
        null
    }

    private fun TransportSkipReason.toTimelineDropReason(): TimelineDropReason = when (this) {
        TransportSkipReason.LATENCY_BUDGET -> TimelineDropReason.LATENCY_BUDGET_SKIP
        TransportSkipReason.EVICTED -> TimelineDropReason.NETWORK_EVICTED
        TransportSkipReason.COVERED -> TimelineDropReason.COVERED
        TransportSkipReason.REWIND -> TimelineDropReason.REWIND
        TransportSkipReason.MISSING_SEQUENCE -> TimelineDropReason.MISSING_SEQUENCE
    }

    private companion object {
        const val NANOS_PER_MICROSECOND = 1_000L
    }
}

File: ios/Sources/MoQKit/Subscribe/internal/pipeline/TrackTimeline.swift
import Foundation

enum TimelineDropReason: Equatable {
    case staleVsPlayback
}

enum TimelineResetReason: Equatable {
    case publisherRewind
    case timestampGap
    case downstreamRecovery
}

enum TimelineDecision<Payload> {
    case admit(PipelineFrame<Payload>)
    case drop(reason: TimelineDropReason, frame: PipelineFrame<Payload>)
    case reset(
        reason: TimelineResetReason,
        epoch: UInt64,
        resumeFrom: PipelineFrame<Payload>?,
        gapUs: UInt64?
    )
}

/// Sole authority for one track's epoch, live edge, freshness, and timestamp-gap resets.
final class TrackTimeline: @unchecked Sendable {
    private let lock = UnfairLock()
    private let policy: TimelinePolicy
    private let timeSource: any PipelineTimeSource

    private var liveEdgeOffsetUs: Int64?
    private var playbackPositionUs: Int64?
    private var lastTimestampUs: Int64?
    private var epoch: UInt64?
    private var latencyTargetUs: Int64

    init(
        policy: TimelinePolicy = PipelinePolicies.timeline,
        timeSource: any PipelineTimeSource = MonotonicPipelineTimeSource()
    ) {
        self.policy = policy
        self.timeSource = timeSource
        self.latencyTargetUs = policy.targetLatencyUs
    }

    func onFrame<Payload>(
        _ frame: PipelineFrame<Payload>,
        arrivalNanos: UInt64? = nil
    ) -> TimelineDecision<Payload> {
        lock.withLock {
            let arrival = arrivalNanos ?? timeSource.nowNanos

            if let epoch, frame.epoch != epoch {
                self.epoch = frame.epoch
                lastTimestampUs = frame.timestampUs
                playbackPositionUs = nil
                resetLiveEdge(timestampUs: frame.timestampUs, arrivalNanos: arrival)
                return .reset(
                    reason: .publisherRewind,
                    epoch: frame.epoch,
                    resumeFrom: frame,
                    gapUs: nil
                )
            }
            if epoch == nil {
                epoch = frame.epoch
            }

            if let previous = lastTimestampUs {
                let gap = Self.absoluteDifference(previous, frame.timestampUs)
                if gap > UInt64(policy.maxGapUs) {
                    lastTimestampUs = frame.timestampUs
                    playbackPositionUs = nil
                    resetLiveEdge(timestampUs: frame.timestampUs, arrivalNanos: arrival)
                    return .reset(
                        reason: .timestampGap,
                        epoch: frame.epoch,
                        resumeFrom: frame,
                        gapUs: gap
                    )
                }
            }

            if let playbackPositionUs,
               frame.timestampUs < playbackPositionUs,
               playbackPositionUs - frame.timestampUs > policy.freshnessBudgetUs
            {
                return .drop(reason: .staleVsPlayback, frame: frame)
            }

            lastTimestampUs = frame.timestampUs
            recordLiveEdge(timestampUs: frame.timestampUs, arrivalNanos: arrival)
            return .admit(frame)
        }
    }

    func onPlaybackPosition(_ positionUs: Int64) {
        precondition(positionUs >= 0, "positionUs must be non-negative")
        lock.withLock {
            playbackPositionUs = positionUs
        }
    }

    func requestReset<Payload>(
        reason: TimelineResetReason = .downstreamRecovery
    ) -> TimelineDecision<Payload> {
        lock.withLock {
            lastTimestampUs = nil
            liveEdgeOffsetUs = nil
            playbackPositionUs = nil
            return .reset(
                reason: reason,
                epoch: epoch ?? 0,
                resumeFrom: nil,
                gapUs: nil
            )
        }
    }

    var currentEpoch: UInt64? {
        lock.withLock { epoch }
    }

    var targetLatencyUs: Int64 {
        lock.withLock { latencyTargetUs }
    }

    func setTargetLatencyUs(_ targetLatencyUs: Int64) {
        precondition(targetLatencyUs >= 0, "target latency must be non-negative")
        lock.withLock {
            latencyTargetUs = targetLatencyUs
        }
    }

    func liveEdgeUs() -> Int64? {
        lock.withLock {
            guard let liveEdgeOffsetUs else { return nil }
            let nowUs = Int64(clamping: timeSource.nowNanos / 1_000)
            let result = nowUs.addingReportingOverflow(liveEdgeOffsetUs)
            return result.overflow ? nil : result.partialValue
        }
    }

    func targetPlaybackUs() -> Int64? {
        guard let liveEdge = liveEdgeUs() else { return nil }
        return max(0, liveEdge - targetLatencyUs)
    }

    func currentLatencyUs() -> Int64? {
        lock.withLock {
            guard let liveEdgeOffsetUs, let playbackPositionUs else { return nil }
            let nowUs = Int64(clamping: timeSource.nowNanos / 1_000)
            let edgeResult = nowUs.addingReportingOverflow(liveEdgeOffsetUs)
            guard !edgeResult.overflow else { return nil }
            return max(0, edgeResult.partialValue - playbackPositionUs)
        }
    }

    private func recordLiveEdge(timestampUs: Int64, arrivalNanos: UInt64) {
        let arrivalUs = Int64(clamping: arrivalNanos / 1_000)
        let result = timestampUs.subtractingReportingOverflow(arrivalUs)
        guard !result.overflow else { return }
        liveEdgeOffsetUs = max(liveEdgeOffsetUs ?? Int64.min, result.partialValue)
    }

    private func resetLiveEdge(timestampUs: Int64, arrivalNanos: UInt64) {
        liveEdgeOffsetUs = nil
        recordLiveEdge(timestampUs: timestampUs, arrivalNanos: arrivalNanos)
    }

    private static func absoluteDifference(_ left: Int64, _ right: Int64) -> UInt64 {
        if left >= right {
            return UInt64(left) - UInt64(right)
        }
        return UInt64(right) - UInt64(left)
    }
}
